teps/0002-custom-tasks.md

```diff
@@ -0,0 +1,392 @@
+
+### Parameter Passing
+
+Custom Task authors should support parameter passing, by supporting a `.spec.params` field (of type [`[]Param`](https://godoc.org/github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1#Param)), and by resolving any `$(params.foo)` placeholders in the CRD type when a `Run` of that type is first reconciled -- this functionality should be implemented by a Go package provided by Tekton, which should be the same one that Tekton itself uses when resolving placeholders in `TaskRun`s and `PipelineRun`s.
+
+This Pipeline pipes its input param `"pl-wait-duration"` to the `example-task`:
+
+```
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: example
+spec:
+  params:
+  - name: pipeline-param
+    default: hello
+  tasks:
+  - name: example-task
+    taskRef:
+      apiVersion: example.dev/v0
+      kind: Example
+      name: my-example
+    params:
+    - name: task-param
+      value: "$(params.pipeline-param)"
+```
+
+In this case, we would expect the `Example` CRD type accepts a parameter named `task-params`, expressed by having a field `.spec.params`, conformant to the `ParamSpec` type.
+
+```
+apiVersion: example.dev/v0
+kind: Example
+metadata:
+  name: my-example
+spec:
+  params:
+  - name: task-param
+    default: goodbye
+  someString: "$(params.task-param), old friend."
+```
+
+When run, the `PipelineRun` controller will create a `Run<Example>`, passing in either the `Pipeline`'s parameter default (`hello`) or the `PipelineRun`-specified override value, if present:
+
+```
+apiVersion: tekton.dev/v1alpha1
+kind: Run
+metadata:
+  name: run-<tekton-generated-suffix>
+spec:
+  ref:
+    apiVersion: example.dev/v0
+    kind: Example
+    name: my-example
+  params:
+  - name: task-param
+    value: hello
+```
+
+The Custom Task author's controller will be watching for creations of `Run`s of this type, and can invoke a Tekton-provided Go package to resolve parameters in string-typed fields, using reflection. This will give them a struct value like:
+
+```
+{
+  "spec": {
+    "params": [ ... ],
+    "someString": "hello, old friend"
+  }
+}
+```
+
+### Result Reporting
+
+Custom Task controllers can report any results via the `Run`'s `.status.results` field:
+
+```
+apiVersion: tekton.dev/v1alpha1
+kind: Run
+metadata:
+  name: run-blah-blah
+spec:
+  ...
+status:
+  results:
+  - name: first-name
+    value: Bob
+```
+
+If their custom defined CRD type has a `.results` field of a compatible type, `Pipeline` validation can take that into account when validating inputs and outputs between Tasks (custom or traditional) in a `Pipeline`.
+
+Controllers can report any results, regardless of whether the underlying CRD object declared them.
+
+### Cancellation
+
+To support cancellation of `Run`s, when a `PipelineRun` is cancelled, the `PipelineRun`  controller will attempt to update any ongoing `Run`s' `.spec.status` to `"Cancelled"`, and update `.status.conditions` to signal unsuccessful execution.
```

Comment:
Does this mean that the `PipelineRun` controller will set a condition on he `Run` resource?
I think setting `Cancelled` to the spec should be enough, and it would be responsibility of the custom task controller to set the condition accordingly.

What will be the behaviour in case of cancellation not supported or delayed or failed for the custom task?

The implementation for `TaskRun` is quite optimistic: we apply the patch to the spec of all `TaskRuns`. If that goes well we assume everything is done and set the completion time. 
I wonder if we should wait for the `TaskRun` to actually report that they stopped running - and similarly for custom tasks.

Reply:
Same question as @afrittoli, updating the `.status.conditions` is the "custom task" controller reponsaility isn't it ?
I think it comes from the fact that this cancellation support would be optional. How can a Custom Task give that information ("I support cancellation", or "I do not support cancellation") to the controller ?

Reply:
This is a bit tricky. If a custom task controller doesn't handle cancellation (i.e., doesn't respond in any way when `.spec.status` is set to `Cancelled`), then the PipelineRun should Do The Right Thing, which I think is to set all sub-task `.status`es to `conditions[@type=Succeeded].status=False` and `reason:RunCancelled`.

Maybe we could specify that PipelineRun controller is responsible for all of this, and custom task authors can optionally detect that case and propagate cancellation however that makes sense for them, or we can say that the PipelineRun controller will set `.spec.status=Cancelled` and then wait to see if anything responds, and _then_ force-cancel the Run? That's sort of how initial update timeout would work, but that means more complex logic in the PipelineRun controller and more nodes on the state diagram -- what happens if the PipelineRun controller restarts while it's waiting to see if a cancellation has been acted upon? Gross.

Ideas welcome. I don't think this specific question should block overall approval of the proposal though, since it only affects a fairly narrow interaction between `Run`s and `PipelineRun`s.

Reply:
Also, we can get this TEP in with still some element to be discussed :wink: